From c1941c08d80852abd8272635b972fa6ce582127b Mon Sep 17 00:00:00 2001
From: Wei-Chiu Chuang <weichiu@cloudera.com>
Date: Mon, 2 May 2016 06:43:25 -0700
Subject: [PATCH 1475/2748] HDFS-10239. Fsshell mv fails if port usage doesn't
 match in src and destination paths. Contributed
 by Kuhu Shukla.

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/TestDFSShell.java

(cherry picked from commit ef3da823573cbf16fd1d84479330dd457f95e0ff)
(cherry picked from commit 9608601330520990d4470971389d574cfa09736e)

Change-Id: Id945f39b42b131dc56b16d62fda71c27b9fbdcb9
---
 .../org/apache/hadoop/fs/shell/MoveCommands.java   |    6 +++-
 .../java/org/apache/hadoop/hdfs/TestDFSShell.java  |   33 +++++++++++++++++++-
 2 files changed, 37 insertions(+), 2 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/shell/MoveCommands.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/shell/MoveCommands.java
index 1c7316a..20cecb4 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/shell/MoveCommands.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/shell/MoveCommands.java
@@ -100,7 +100,11 @@ protected void processOptions(LinkedList<String> args) throws IOException {
 
     @Override
     protected void processPath(PathData src, PathData target) throws IOException {
-      if (!src.fs.getUri().equals(target.fs.getUri())) {
+      String srcUri = src.fs.getUri().getScheme() + "://" +
+          src.fs.getUri().getHost();
+      String dstUri = target.fs.getUri().getScheme() + "://" +
+          target.fs.getUri().getHost();
+      if (!srcUri.equals(dstUri)) {
         throw new PathIOException(src.toString(),
             "Does not match target filesystem");
       }
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/TestDFSShell.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/TestDFSShell.java
index 02acf01..3e031e5 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/TestDFSShell.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/TestDFSShell.java
@@ -527,7 +527,38 @@ public void testErrOutPut() throws Exception {
       }
     }
   }
-  
+
+  @Test
+  public void testMoveWithTargetPortEmpty() throws Exception {
+    Configuration conf = new HdfsConfiguration();
+    MiniDFSCluster cluster = null;
+    try {
+      cluster = new MiniDFSCluster.Builder(conf)
+          .format(true)
+          .numDataNodes(2)
+          .nameNodePort(8020)
+          .waitSafeMode(true)
+          .build();
+      FileSystem srcFs = cluster.getFileSystem();
+      FsShell shell = new FsShell();
+      shell.setConf(conf);
+      String[] argv = new String[2];
+      argv[0] = "-mkdir";
+      argv[1] = "/testfile";
+      ToolRunner.run(shell, argv);
+      argv = new String[3];
+      argv[0] = "-mv";
+      argv[1] = srcFs.getUri() + "/testfile";
+      argv[2] = "hdfs://localhost/testfile2";
+      int ret = ToolRunner.run(shell, argv);
+      assertEquals("mv should have succeeded", 0, ret);
+    } finally {
+      if (cluster != null) {
+        cluster.shutdown();
+      }
+    }
+  }
+
   @Test (timeout = 30000)
   public void testURIPaths() throws Exception {
     Configuration srcConf = new HdfsConfiguration();
-- 
1.7.9.5

